### PR TITLE
A few changes that made mutate_cpp working on Ubuntu 16.04

### PR DESCRIPTION
**Pull request by igormironchik**
Now it's simple to initialize and run mutate_cpp in the repository, just clone and do the job. Patches made to work on Ubuntu 16.04. But patches now use \r\n line endings, this is default behavior for git?!

### CODE REVIEW

**Comment by nlohmann** [comment]
Thanks for the PR. I shall have a look at it soon. I hope that the project was still useful.

**Comment by igormironchik** [comment]
mutate_cpp is really interesting but usual C++ file mutates approximately 700 times. This is 700 rebuilds for one file (if code places in header). In a small project on usual hardware this is a working day. That is why I see mutate_cpp as a toy for developers... But it can help... If a developer has enough time or a good hardware.

File: README.md [hunk 1]
```diff
@@ -34,8 +34,8 @@ After checking out the sources from this repository, you need to create a
 [virtual environment](https://docs.python.org/3/tutorial/venv.html) and install the required packages:
 
 ```bash
-virtualenv -p python3 venv
-venv/bin/pip install -r requirements
+python3 -m venv venv
```
Comment (nlohmann) [inline README.md]: This assumes that the module `venv` is installed, right? I tried this on a Ubuntu system today, and `virtualemnv` worked, whereas `python -m venv` did not.

File: app/__init__.py [hunk 2]
```diff
@@ -4,6 +4,10 @@
 from flask_compress import Compress
 from flask_sqlalchemy import SQLAlchemy
 from flask_humanize import Humanize
+import os.path
+import sys
+
+sys.path.append( os.path.dirname( os.path.abspath( __file__ ) ) )
```
Comment (nlohmann) [inline app/__init__.py]: I never needed that so far. Is this to make the code robust against the working directory?

File: app/utils/SourceFile.py [hunk 3]
```diff
@@ -117,14 +117,14 @@ def __create_patch(self, line_number: int, replacement: Replacement) -> str:
         ))
 
         # lines: context before
-        patch_lines += [' ' + x + '\n' for x in context_before]
+        patch_lines += [' ' + x + "\r\n" for x in context_before]
```
Comment (nlohmann) [inline app/utils/SourceFile.py]: Why is `\r\n` needed here?

File: README.md (hunk 1 above)
Comment (igormironchik) [reply README.md]: I have opposite situation. ```virtualenv -p python3 venv``` doesn't work for me on Ubuntu 16.04 whereas ```python3 -m venv venv``` is working. And on page https://docs.python.org/3/tutorial/venv.html you can see example:

"To create a virtual environment, decide upon a directory where you want to place it, and run the venv module as a script with the directory path"

```python3 -m venv tutorial-env```

File: app/__init__.py (hunk 2 above)
Comment (igormironchik) [reply app/__init__.py]: Yes, this line is for ```create_db.py``` to be working when launched exactly from the root directory of the repository. I had problems with ```from config import ...``` Python couldn't find module ```config```. And when I added this line to ```app/__init__.py ```create_db.py``` became working.

File: app/utils/SourceFile.py (hunk 3 above)
Comment (igormironchik) [reply app/utils/SourceFile.py]: Git can clone files with different line endings mode. By default, as I guess, git clone with ```\r\n``` and pushes with ```\n```. So on my Ubuntu 16.04 I have repositories with ```\r\n``` line endings in files. So patch didn't work with ```\n``` and works with ```\r\n```. But to be fair it's better to handle both situations...

File: app/__init__.py (hunk 2 above)
Comment (igormironchik) [reply app/__init__.py]: Sure, user can tune python's environment so all will work without this line. But I'm not sure about ```from config import ....```, why not ```from app.config import ....```?

File: README.md (hunk 1 above)
Comment (nlohmann) [reply README.md]: Ok, then there is apparently more than 1 way to do this...

Comment (nlohmann) [reply README.md]: Ok. Please leave this as it is. Maybe add a comment below that some systems require `python3 -m venv venv` to create a virtual environment.

File: app/utils/Executor.py [hunk 4]
```diff
@@ -74,33 +74,37 @@ def killer(process):
...
+                      self.__apply_command(patch, 'quickcheck_command') and \
+                      self.__apply_command(patch, 'test_command')
+
+            if success:
+                 patch.state = 'survived'
+                 db.session.commit()
+
+            # step 4: revert patch
+            self.__execute_command_timeout('patch -p1 --reverse --input={patchfile} {inputfile}'.format(patchfile=patchfile.name, inputfile=file.filename),
                                        cwd='/')
```
Comment (nlohmann) [inline app/utils/Executor.py]: What is the rationale for this changed block?

File: db_create.py [hunk 5]
```diff
@@ -1,8 +1,8 @@
 # coding=utf-8
 
 from migrate.versioning import api
-from config import SQLALCHEMY_DATABASE_URI
-from config import SQLALCHEMY_MIGRATE_REPO
+from app.config import SQLALCHEMY_DATABASE_URI
+from app.config import SQLALCHEMY_MIGRATE_REPO
 from app import db
```
Comment (nlohmann) [inline db_create.py]: Wouldn't these changes also be required in `db_migrate.py` and `db_upgrade.py`?

File: app/utils/Executor.py (hunk 4 above)
Comment (igormironchik) [reply app/utils/Executor.py]: This is needed to work properly on Ubuntu 16.04. If launch ```patch -p1 --reverse --input={patchfile}``` then patch will fail and request file name to patch. And with ```patch -p1 --reverse --input={patchfile} {inputfile}``` all works as expected.

File: db_create.py (hunk 5 above)
Comment (igormironchik) [reply db_create.py]: I think so. I didn't use these scripts, so didn't make any changes to them.

**Comment by igormironchik** [comment]
I restored README.md and replaced ```from config import ...``` with ```from app.config import ...``` in db scripts. And I guess that PR #2 is correct because ```venv/bin/pip install -r requirements``` seems to need ```-r requirements.txt```

**Comment by nlohmann** [comment]
Thanks a lot!

**Comment by igormironchik** [comment]
Could you run ```patch -p1 --input=/var/folders/sh/026s5d0s7p32_ytprmw0lyx00000gn/T/tmpwcvbyykd /Users/***/cmake-example/src/example.cpp``` by hands? What does it say? I guess that this is problem of line endings in patch. I guess that this is impossible to generate a patch in Python that will handle all possible situations because in Python line endings always converts to ```\n```. I thought about it and didn't find a way to solve the problem of line endings in Python when reading text.

**Comment by igormironchik** [comment]
I caught exactly the same problem on my Ubuntu on that days, that is why I created a PR that solved the problem for me. You can edit ```mutate_cpp/app/utils/SourceFile.py```

And change ```\r\n``` to ```\n```

```
# lines: context before
patch_lines += [' ' + x + "\r\n" for x in context_before]
# line: the old value
patch_lines.append('-' + old_line + "\r\n")
# line: the new value
if replacement.new_val is not None:
    patch_lines.append('+' + new_line + "\r\n")
# lines: context after
patch_lines += [' ' + x + "\r\n" for x in context_after]
```

**Comment by igormironchik** [comment]
I can suggest to refuse from ```patch``` at all, and implement patching and reverting in Python, but again line endings can be messed...

**Pull request merged**